Be 03/04/07/feat(forget password) forgot, set password, and migration

Kocannn opened the pull request:
- add migration + seeding with goose combine with cobra command
usage migrate
- go run main.go migrate:create name_migration // for create new file migration
- go run main.go migrate:up // for up migration
- go run main.go migrate:down // for down migration
usage for seeding 
- go run main.go seed:create name_seed // for create new file seed
- go run main.go seed:up // run seeding file

- add feature to reset password

Kocannn committed:
feat(auth): add forgot and reset password functionality

- Implement forgot password handler to generate reset token and link
- Add reset password form to validate and update user password
- Create HTML templates for reset password and email confirmation
- Update routes to include forgot and reset password endpoints
- Add usecase methods for token verification and password reset

Kocannn committed:
feat(reset-password): implement reset password functionality

Add reset password functionality across repository, usecase, and handler
layers. This includes:

- Repository method to update user password in the database.
- Usecase logic to verify token, hash password, and call repository.
- HTTP handler to handle reset password requests.

Also updated the domain model to include token and password fields in
ForgotPassword struct and added the necessary interface methods.

BREAKING CHANGE: The reset password endpoint now requires a token and
password in the request body.

Kocannn committed:
feat(users): integrate reset password link in email template

- Updated `ForgotPassword` use case to pass reset link to the email
template.
- Fixed email template rendering using `html/template` package.
- Adjusted reset link to point to the correct frontend URL.
- Improved error handling for email sending process.

This ensures users receive a functional reset password link in their
email.

Kocannn committed:
feat(cmd): add migration commands with goose integration

Added new migration commands `migrate:create`, `migrate:up`, and
`migrate:down` using the goose library. Updated dependencies for
cobra and pflag to their latest versions. Removed unused imports
and legacy migration code.

BREAKING CHANGE: The `dbMigrate` command has been replaced with
new migration commands.

Kocannn committed:
feat(auth): enhance password reset and seeding functionality

- Updated `ForgotPassword` to use frontend base URL for reset link.
- Added email and password validation in `ResetPassword` handler.
- Introduced `ConfirmPassword` field in `ForgotPassword` struct.
- Removed unused `ShowResetPasswordForm` and related assets.
- Added seeding commands (`seed:create`, `seed:up`, `seed:down`).
- Modified token generation to accept custom expiration time.
- Updated migration command to include table prefix.

Kocannn committed:
refactor(auth): optimize ResetPassword logic and queries

- Fixed token validation logic to correctly handle used tokens.
- Improved GORM queries by using pointer semantics for better
performance.
- Enhanced error logging with `WithError` for better debugging context.
- Removed redundant `Model()` call in the initial token query.
- Improved code readability with named variables and consistent
formatting.

These changes enhance code clarity, maintainability, and runtime
efficiency.

Kocannn committed:
feat(cmd, database): enhance migration and seeding commands

- Updated `cmd.go` to remove unused `seedDown` command.
- Adjusted `migrate.go` commands to align with `goose` CLI changes.
- Added new database migration files for `users`, `logout`, `events`,
  `registration_events`, `event_pays`, `event_speakers`, `event_tags`,
  `images`, and `testing` tables.
- Ensured proper foreign key constraints and sequence handling.

These changes improve database schema management and command usability.

AfandyW commented on app/users/usecase/forgot_password.go:
```diff
@@ -0,0 +1,84 @@
+	"github.com/sirupsen/logrus"
+)
+
+func (us *usecase) ForgotPassword(ctx context.Context, emailForgot domain.ForgotPassword) (user domain.User, resetLink string, err error) {
```
mungkin remove saja return user dan reset link. karna nnti akan di kirim ke email buat informasi reset link. return user kalau tidak digunakan, di remove

AfandyW commented on app/users/usecase/forgot_password.go:
```diff
@@ -0,0 +1,84 @@
+		return
+	}
+
+	resetToken, err := us.jwt.GenerateAccessToken(ctx, &user, 5)
```
reset token kasih waktu 30 menitan, siapa tau ada issue jaringan dia ngga bisa buka email atau emailnya lumayan lama masuk.

AfandyW commented on pkg/jwt/generate_token.go:
```diff
@@ -8,8 +8,8 @@ import (
 
-func (j *jwtConfig) GenerateAccessToken(c context.Context, user *domain.User) (*string, error) {
-	expiredTime := time.Now().Local().Add(60 * time.Minute)
+func (j *jwtConfig) GenerateAccessToken(c context.Context, user *domain.User, durationToken int) (*string, error) {
```
tambahkan comment di atas fucntion ini. biar kalau di hover functionnya ketahuan detail waktu duration token.
misal
durationToken in minute

AfandyW commented on pkg/jwt/jwt.go:
```diff
@@ -21,7 +21,7 @@ type (
 
 	JWT interface {
-		GenerateAccessToken(c context.Context, user *domain.User) (*string, error)
+		GenerateAccessToken(c context.Context, user *domain.User, expiredTime int) (*string, error)
```
disini juga taruh penjelasan expired time format apa (minute/hour)

AfandyW commented on app/users/delivery/http/forgot_password_user.go:
```diff
@@ -0,0 +1,49 @@
+		return
+	}
+
+	_, _, err = h.usecase.ForgotPassword(r.Context(), forgotPassword)
```
kalau banyak yang di igrnore, mending di efisiensikan

AfandyW commented on app/users/delivery/http/reset_password.go:
```diff
@@ -0,0 +1,74 @@
+	if err := json.Unmarshal(bodyBytes, &forgotPasswordInstance); err != nil {
+		utils.Response(domain.HttpResponse{
+			Code:    500,
+			Message: err.Error(),
```
jangan langsung kasih error dari code program. mungkin bisa di kasih keterangan failed to unmarshal

AfandyW commented on app/users/delivery/http/reset_password.go:
```diff
@@ -0,0 +1,74 @@
+	if err := h.usecase.ResetPassword(r.Context(), forgotPasswordInstance); err != nil {
+		utils.Response(domain.HttpResponse{
+			Code:    500,
+			Message: err.Error(),
```
pastikan message errornya jelas

AfandyW commented on app/users/repository/forgot_password.go:
```diff
@@ -0,0 +1,23 @@
+	err = repo.db.DB(ctx).Create(&domain.ResetPasswordToken{
+		Token:      token,
+		UserID:     uint64(user.ID),
+		ExpiryDate: expired_at,
```
nama variable kalau bisa kasih camelCase, biar seragam

AfandyW commented on app/users/repository/reset_password.go:
```diff
@@ -0,0 +1,39 @@
+
+func (repo *repository) ResetPassword(ctx context.Context, email, password, token string) error {
+	if err := repo.db.StartTransaction(ctx, func(ctx context.Context) error {
+		ResetPasswordTokenInstance := domain.ResetPasswordToken{}
```
hindari penggunaan huruf besar di golang, karna ke hitunga global variable, kalau bisa pakai format camelcase. kecuali variable itu mau di import atau di gunakan di function lain

AfandyW commented on cmd/serve_http.go:
```diff
@@ -125,6 +125,8 @@ func registerHandler(app app.App) *mux.Router {
 	auth.HandleFunc("/register", app.UserHandler.Register).Methods(http.MethodPost)
 	auth.HandleFunc("/login", app.UserHandler.Login).Methods(http.MethodPost)
+	auth.HandleFunc("/forgot_password", app.UserHandler.ForgotPassword).Methods(http.MethodPost)
+	auth.HandleFunc("/set_password", app.UserHandler.ResetPassword).Methods(http.MethodPost)
```
kasih method put, karna update data

AfandyW commented on config/config.go:
```diff
@@ -80,6 +81,7 @@ func GetConfig() Config {
 			CORS_ALLOWED_METHODS: methods,
 			CORS_ALLOWED_HEADERS: headers,
 			BaseURL:              viper.GetString("BASE_URL"),
```
refactor aja kalau nemu format berbeda haha

AfandyW commented on domain/user.go:
```diff
@@ -86,6 +92,13 @@ type (
 	}
+
+	ForgotPassword struct {
+		Email           string `json:"email,omitempty"`
```
apakah ini harus required?

Kocannn replied:
soalnya untuk struct ForgotPassowrdnya masih saya pake buat ba set passwordnya yang cuma butuh token, password sama confirm passwordnya dari struct nya jadi sa taruh omitempty saja, tapi untuk forgot passwordnya sudh saya taruh validator untuk ba atasi kalo email yang dia masukkan kosong

Kocannn committed:
feat(auth): improve forgot/reset password validation & flow

- Change HTTP methods for forgot_password and set_password endpoints to
PUT
- Add email format validation using regex in ForgotPassword handler
- Refactor ForgotPassword usecase and repository to return only error
- Improve error messages for request body read/unmarshal failures
- Fix password validation logic and error message in ResetPassword
handler
- Update JWT token generation to use 30 minutes for reset token
- Refactor variable naming for consistency in repository methods
- Update domain interfaces to match new ForgotPassword signature

AfandyW reviewed (approved):
good
